### ISSUE TITLE
trillian_log_server Mysql error: 'Could not start tree TX: Error 1040: Too many connections'

### ISSUE THREAD

**Issue by pgporada** [google/trillian#1142]
Hi,

Much like https://github.com/google/trillian/issues/901 I am running into a 1040 too many connections error. The initial mysql max connections pool was set to 256 connections and that was quickly getting exhausted just from testing. The max connections setting was bumped up and the error went away. What I am curious about is if `packets.go` and `connection.go` should be barfing errors as seen below.
```
    May 17 17:07:12 myHost systemd[1]: Started trillian_log_server.
    May 17 17:07:12 myHost systemd[1]: Starting trillian_log_server...
    May 17 17:07:12 myHost trillian_log_server[13783]: I0517 17:07:12.067135   13783 mysql_quota_provider.go:44] Using MySQL QuotaManager
    May 17 17:07:12 myHost trillian_log_server[13783]: I0517 17:07:12.068549   13783 main.go:160] RPC server starting on 0.0.0.0:8090
    May 17 17:07:12 myHost trillian_log_server[13783]: I0517 17:07:12.068857   13783 main.go:169] Deleted tree GC started
    May 17 17:07:12 myHost trillian_log_server[13783]: I0517 17:07:12.069414   13783 main.go:144] HTTP server starting on 0.0.0.0:8091
    May 17 17:11:47 myHost trillian_log_server[13783]: W0517 17:11:47.600041   13783 tree_storage.go:148] Could not start tree TX: Error 1040: Too many connections
    May 17 17:51:44 myHost trillian_log_server[13783]: W0517 17:51:44.961843   13783 tree_storage.go:148] Could not start tree TX: Error 1040: Too many connections
    May 17 19:31:46 myHost trillian_log_server[13783]: W0517 19:31:46.223943   13783 tree_storage.go:148] Could not start tree TX: Error 1040: Too many connections
    May 18 10:31:56 myHost trillian_log_server[13783]: W0518 10:31:56.813383   13783 tree_storage.go:148] Could not start tree TX: Error 1040: Too many connections
    May 18 10:31:56 myHost trillian_log_server[13783]: W0518 10:31:56.823923   13783 tree_storage.go:148] Could not start tree TX: Error 1040: Too many connections
    May 18 13:02:05 myHost trillian_log_server[13783]: W0518 13:02:05.079123   13783 tree_storage.go:148] Could not start tree TX: Error 1040: Too many connections
    May 18 14:31:40 myHost trillian_log_server[13783]: [mysql] 2018/05/18 14:31:40 packets.go:36: unexpected EOF
    May 18 14:33:00 myHost trillian_log_server[13783]: [mysql] 2018/05/18 14:33:00 packets.go:36: unexpected EOF
    May 18 14:34:40 myHost trillian_log_server[13783]: [mysql] 2018/05/18 14:34:40 connection.go:99: invalid connection
    May 18 14:34:40 myHost trillian_log_server[13783]: [mysql] 2018/05/18 14:34:40 connection.go:99: invalid connection
    May 18 14:36:00 myHost trillian_log_server[13783]: [mysql] 2018/05/18 14:36:00 packets.go:36: unexpected EOF
    May 18 14:37:40 myHost trillian_log_server[13783]: [mysql] 2018/05/18 14:37:40 connection.go:99: invalid connection
    May 18 14:39:00 myHost trillian_log_server[13783]: [mysql] 2018/05/18 14:39:00 packets.go:36: unexpected EOF
    May 18 14:40:40 myHost trillian_log_server[13783]: [mysql] 2018/05/18 14:40:40 connection.go:99: invalid connection
    May 18 14:42:00 myHost trillian_log_server[13783]: [mysql] 2018/05/18 14:42:00 packets.go:36: unexpected EOF
    May 18 14:43:40 myHost trillian_log_server[13783]: [mysql] 2018/05/18 14:43:40 connection.go:99: invalid connection
    May 18 14:45:00 myHost trillian_log_server[13783]: [mysql] 2018/05/18 14:45:00 packets.go:36: unexpected EOF
    May 18 14:46:40 myHost trillian_log_server[13783]: [mysql] 2018/05/18 14:46:40 connection.go:99: invalid connection
    May 18 14:48:00 myHost trillian_log_server[13783]: [mysql] 2018/05/18 14:48:00 packets.go:36: unexpected EOF
    May 18 14:49:40 myHost trillian_log_server[13783]: [mysql] 2018/05/18 14:49:40 connection.go:99: invalid connection
    May 18 14:51:00 myHost trillian_log_server[13783]: [mysql] 2018/05/18 14:51:00 packets.go:36: unexpected EOF
    May 18 14:52:40 myHost trillian_log_server[13783]: [mysql] 2018/05/18 14:52:40 connection.go:99: invalid connection
    May 18 14:54:00 myHost trillian_log_server[13783]: [mysql] 2018/05/18 14:54:00 packets.go:36: unexpected EOF
    May 18 14:55:40 myHost trillian_log_server[13783]: [mysql] 2018/05/18 14:55:40 connection.go:99: invalid connection
    May 18 14:57:00 myHost trillian_log_server[13783]: [mysql] 2018/05/18 14:57:00 packets.go:36: unexpected EOF
    May 18 14:58:40 myHost trillian_log_server[13783]: [mysql] 2018/05/18 14:58:40 connection.go:99: invalid connection
    May 18 15:00:00 myHost trillian_log_server[13783]: [mysql] 2018/05/18 15:00:00 packets.go:36: unexpected EOF
    May 18 15:01:40 myHost trillian_log_server[13783]: [mysql] 2018/05/18 15:01:40 connection.go:99: invalid connection
    May 18 15:03:00 myHost trillian_log_server[13783]: [mysql] 2018/05/18 15:03:00 packets.go:36: unexpected EOF
    May 18 15:04:40 myHost trillian_log_server[13783]: [mysql] 2018/05/18 15:04:40 connection.go:99: invalid connection
    May 18 15:06:00 myHost trillian_log_server[13783]: [mysql] 2018/05/18 15:06:00 packets.go:36: unexpected EOF
    May 18 15:07:40 myHost trillian_log_server[13783]: [mysql] 2018/05/18 15:07:40 connection.go:99: invalid connection
    May 18 15:09:00 myHost trillian_log_server[13783]: [mysql] 2018/05/18 15:09:00 packets.go:36: unexpected EOF
    May 18 15:10:40 myHost trillian_log_server[13783]: [mysql] 2018/05/18 15:10:40 connection.go:99: invalid connection
    May 18 15:12:00 myHost trillian_log_server[13783]: [mysql] 2018/05/18 15:12:00 packets.go:36: unexpected EOF
    May 18 15:13:40 myHost trillian_log_server[13783]: [mysql] 2018/05/18 15:13:40 connection.go:99: invalid connection
    May 18 15:15:00 myHost trillian_log_server[13783]: [mysql] 2018/05/18 15:15:00 packets.go:36: unexpected EOF
    May 18 15:16:40 myHost trillian_log_server[13783]: [mysql] 2018/05/18 15:16:40 connection.go:99: invalid connection
    May 18 15:18:00 myHost trillian_log_server[13783]: [mysql] 2018/05/18 15:18:00 packets.go:36: unexpected EOF
    May 18 15:19:40 myHost trillian_log_server[13783]: [mysql] 2018/05/18 15:19:40 connection.go:99: invalid connection
    May 18 15:52:05 myHost trillian_log_server[13783]: W0518 15:52:05.034577   13783 tree_storage.go:148] Could not start tree TX: Error 1040: Too many connections
    May 18 16:12:06 myHost trillian_log_server[13783]: W0518 16:12:06.388961   13783 tree_storage.go:148] Could not start tree TX: Error 1040: Too many connections
```

**Comment by Martin2112** [comment]
#901 is specific to the way the map server currently uses transactions to apply updates. If you have a log and map server accessing the same database server then you'll need to increase its max connections above 256.

If you only have a log server using the database and nothing else is tying up connections then this is something different and we'll probably need to collect more information. It shouldn't be holding on to a lot of connections.

**Comment by pgporada** [comment]
I only have the log server using the database.

**Comment by jsha** [comment]
Part of the issue seems to be that go-sql-driver doesn't recover gracefully when it gets an `Error 1040` from the MySQL server. I still need to produce a rigorous repro case to show that, but that seems to be the behavior we're observing. As a first-line fix, I think it would be great to provide a `--max_db_conns` flag that would call `db.SetMaxOpenConns` on the corresponding database.

**Comment by RJPercival** [comment]
PR #1288 has added a `--mysql_max_conns` flag, so now the maximum number of connections that will be made from a Trillian binary to MySQL can be limited.

**Comment by RJPercival** [comment]
@jsha: Were you able to produce a repro case?

**Comment by jsha** [comment]
I haven't spent the time to isolate a repro case, but I think it repros fairly easy with an out-of-the-box MySQL, which has a [default server-side limit of 151 connections](https://dev.mysql.com/doc/refman/8.0/en/server-system-variables.html#sysvar_max_connections). If you have enough traffic that there are more that 151 connections open at a time you'll run into this.

I think the specific issue here was that the log_signer may not recover properly if it gets an error 1040. However, I realize that's pretty vague. I'm happy to consider this closed by #1288 for now, and if I find the time later to reproduce the "log_signer doesn't recover" problem, I'll file a more detailed issue.
